File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/FinalCalculationPeriodDateAdjustment.cpp
// FinalCalculationPeriodDateAdjustment.cpp 
#include "FinalCalculationPeriodDateAdjustment.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FinalCalculationPeriodDateAdjustment::FinalCalculationPeriodDateAdjustment(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //relevantUnderlyingDateReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relevantUnderlyingDateReferenceNode = xmlNode->FirstChildElement("relevantUnderlyingDateReference");

   if(relevantUnderlyingDateReferenceNode){relevantUnderlyingDateReferenceIsNull_ = false;}
   else{relevantUnderlyingDateReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relevantUnderlyingDateReferenceNode , address : " << relevantUnderlyingDateReferenceNode << std::endl;
   #endif
   if(relevantUnderlyingDateReferenceNode)
   {
      if(relevantUnderlyingDateReferenceNode->Attribute("href") || relevantUnderlyingDateReferenceNode->Attribute("id"))
      {
          if(relevantUnderlyingDateReferenceNode->Attribute("id"))
          {
             relevantUnderlyingDateReferenceIDRef_ = relevantUnderlyingDateReferenceNode->Attribute("id");
             relevantUnderlyingDateReference_ = boost::shared_ptr<RelevantUnderlyingDateReference>(new RelevantUnderlyingDateReference(relevantUnderlyingDateReferenceNode));
             relevantUnderlyingDateReference_->setID(relevantUnderlyingDateReferenceIDRef_);
             IDManager::instance().setID(relevantUnderlyingDateReferenceIDRef_,relevantUnderlyingDateReference_);
          }
          else if(relevantUnderlyingDateReferenceNode->Attribute("href")) { relevantUnderlyingDateReferenceIDRef_ = relevantUnderlyingDateReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relevantUnderlyingDateReference_ = boost::shared_ptr<RelevantUnderlyingDateReference>(new RelevantUnderlyingDateReference(relevantUnderlyingDateReferenceNode));}
   }

   //swapStreamReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* swapStreamReferenceNode = xmlNode->FirstChildElement("swapStreamReference");

   if(swapStreamReferenceNode){swapStreamReferenceIsNull_ = false;}
   else{swapStreamReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- swapStreamReferenceNode , address : " << swapStreamReferenceNode << std::endl;
   #endif
   if(swapStreamReferenceNode)
   {
      if(swapStreamReferenceNode->Attribute("href") || swapStreamReferenceNode->Attribute("id"))
      {
          if(swapStreamReferenceNode->Attribute("id"))
          {
             swapStreamReferenceIDRef_ = swapStreamReferenceNode->Attribute("id");
             swapStreamReference_ = boost::shared_ptr<InterestRateStreamReference>(new InterestRateStreamReference(swapStreamReferenceNode));
             swapStreamReference_->setID(swapStreamReferenceIDRef_);
             IDManager::instance().setID(swapStreamReferenceIDRef_,swapStreamReference_);
          }
          else if(swapStreamReferenceNode->Attribute("href")) { swapStreamReferenceIDRef_ = swapStreamReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { swapStreamReference_ = boost::shared_ptr<InterestRateStreamReference>(new InterestRateStreamReference(swapStreamReferenceNode));}
   }

   //businessDayConventionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessDayConventionNode = xmlNode->FirstChildElement("businessDayConvention");

   if(businessDayConventionNode){businessDayConventionIsNull_ = false;}
   else{businessDayConventionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessDayConventionNode , address : " << businessDayConventionNode << std::endl;
   #endif
   if(businessDayConventionNode)
   {
      if(businessDayConventionNode->Attribute("href") || businessDayConventionNode->Attribute("id"))
      {
          if(businessDayConventionNode->Attribute("id"))
          {
             businessDayConventionIDRef_ = businessDayConventionNode->Attribute("id");
             businessDayConvention_ = boost::shared_ptr<BusinessDayConventionEnum>(new BusinessDayConventionEnum(businessDayConventionNode));
             businessDayConvention_->setID(businessDayConventionIDRef_);
             IDManager::instance().setID(businessDayConventionIDRef_,businessDayConvention_);
          }
          else if(businessDayConventionNode->Attribute("href")) { businessDayConventionIDRef_ = businessDayConventionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessDayConvention_ = boost::shared_ptr<BusinessDayConventionEnum>(new BusinessDayConventionEnum(businessDayConventionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<RelevantUnderlyingDateReference> FinalCalculationPeriodDateAdjustment::getRelevantUnderlyingDateReference()
{
   if(!this->relevantUnderlyingDateReferenceIsNull_){
        if(relevantUnderlyingDateReferenceIDRef_ != ""){
             return boost::shared_static_cast<RelevantUnderlyingDateReference>(IDManager::instance().getID(relevantUnderlyingDateReferenceIDRef_));
        }else{
             return this->relevantUnderlyingDateReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<RelevantUnderlyingDateReference>();
   }
}
boost::shared_ptr<InterestRateStreamReference> FinalCalculationPeriodDateAdjustment::getSwapStreamReference()
{
   if(!this->swapStreamReferenceIsNull_){
        if(swapStreamReferenceIDRef_ != ""){
             return boost::shared_static_cast<InterestRateStreamReference>(IDManager::instance().getID(swapStreamReferenceIDRef_));
        }else{
             return this->swapStreamReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InterestRateStreamReference>();
   }
}
boost::shared_ptr<BusinessDayConventionEnum> FinalCalculationPeriodDateAdjustment::getBusinessDayConvention()
{
   if(!this->businessDayConventionIsNull_){
        if(businessDayConventionIDRef_ != ""){
             return boost::shared_static_cast<BusinessDayConventionEnum>(IDManager::instance().getID(businessDayConventionIDRef_));
        }else{
             return this->businessDayConvention_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessDayConventionEnum>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/FinalCalculationPeriodDateAdjustment.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FinalCalculationPeriodDateAdjustment : ISerialized
    {
        public FinalCalculationPeriodDateAdjustment(XmlNode xmlNode)
        {
            XmlNode relevantUnderlyingDateReferenceNode = xmlNode.SelectSingleNode("relevantUnderlyingDateReference");
            
            if (relevantUnderlyingDateReferenceNode != null)
            {
                if (relevantUnderlyingDateReferenceNode.Attributes["href"] != null || relevantUnderlyingDateReferenceNode.Attributes["id"] != null) 
                {
                    if (relevantUnderlyingDateReferenceNode.Attributes["id"] != null) 
                    {
                        relevantUnderlyingDateReferenceIDRef_ = relevantUnderlyingDateReferenceNode.Attributes["id"].Value;
                        RelevantUnderlyingDateReference ob = new RelevantUnderlyingDateReference(relevantUnderlyingDateReferenceNode);
                        IDManager.SetID(relevantUnderlyingDateReferenceIDRef_, ob);
                    }
                    else if (relevantUnderlyingDateReferenceNode.Attributes["href"] != null)
                    {
                        relevantUnderlyingDateReferenceIDRef_ = relevantUnderlyingDateReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relevantUnderlyingDateReference_ = new RelevantUnderlyingDateReference(relevantUnderlyingDateReferenceNode);
                    }
                }
                else
                {
                    relevantUnderlyingDateReference_ = new RelevantUnderlyingDateReference(relevantUnderlyingDateReferenceNode);
                }
            }
            
        
            XmlNode swapStreamReferenceNode = xmlNode.SelectSingleNode("swapStreamReference");
            
            if (swapStreamReferenceNode != null)
            {
                if (swapStreamReferenceNode.Attributes["href"] != null || swapStreamReferenceNode.Attributes["id"] != null) 
                {
                    if (swapStreamReferenceNode.Attributes["id"] != null) 
                    {
                        swapStreamReferenceIDRef_ = swapStreamReferenceNode.Attributes["id"].Value;
                        InterestRateStreamReference ob = new InterestRateStreamReference(swapStreamReferenceNode);
                        IDManager.SetID(swapStreamReferenceIDRef_, ob);
                    }
                    else if (swapStreamReferenceNode.Attributes["href"] != null)
                    {
                        swapStreamReferenceIDRef_ = swapStreamReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        swapStreamReference_ = new InterestRateStreamReference(swapStreamReferenceNode);
                    }
                }
                else
                {
                    swapStreamReference_ = new InterestRateStreamReference(swapStreamReferenceNode);
                }
            }
            
        
            XmlNode businessDayConventionNode = xmlNode.SelectSingleNode("businessDayConvention");
            
            if (businessDayConventionNode != null)
            {
                if (businessDayConventionNode.Attributes["href"] != null || businessDayConventionNode.Attributes["id"] != null) 
                {
                    if (businessDayConventionNode.Attributes["id"] != null) 
                    {
                        businessDayConventionIDRef_ = businessDayConventionNode.Attributes["id"].Value;
                        BusinessDayConventionEnum ob = new BusinessDayConventionEnum(businessDayConventionNode);
                        IDManager.SetID(businessDayConventionIDRef_, ob);
                    }
                    else if (businessDayConventionNode.Attributes["href"] != null)
                    {
                        businessDayConventionIDRef_ = businessDayConventionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessDayConvention_ = new BusinessDayConventionEnum(businessDayConventionNode);
                    }
                }
                else
                {
                    businessDayConvention_ = new BusinessDayConventionEnum(businessDayConventionNode);
                }
            }
            
        
        }
        
    
        #region RelevantUnderlyingDateReference_
        private RelevantUnderlyingDateReference relevantUnderlyingDateReference_;
        public RelevantUnderlyingDateReference RelevantUnderlyingDateReference_
        {
            get
            {
                if (this.relevantUnderlyingDateReference_ != null)
                {
                    return this.relevantUnderlyingDateReference_; 
                }
                else if (this.relevantUnderlyingDateReferenceIDRef_ != null)
                {
                    relevantUnderlyingDateReference_ = IDManager.getID(relevantUnderlyingDateReferenceIDRef_) as RelevantUnderlyingDateReference;
                    return this.relevantUnderlyingDateReference_; 
                }
                else
                {
                      return this.relevantUnderlyingDateReference_; 
                }
            }
            set
            {
                if (this.relevantUnderlyingDateReference_ != value)
                {
                    this.relevantUnderlyingDateReference_ = value;
                }
            }
        }
        #endregion
        
        public string relevantUnderlyingDateReferenceIDRef_ { get; set; }
        #region SwapStreamReference_
        private InterestRateStreamReference swapStreamReference_;
        public InterestRateStreamReference SwapStreamReference_
        {
            get
            {
                if (this.swapStreamReference_ != null)
                {
                    return this.swapStreamReference_; 
                }
                else if (this.swapStreamReferenceIDRef_ != null)
                {
                    swapStreamReference_ = IDManager.getID(swapStreamReferenceIDRef_) as InterestRateStreamReference;
                    return this.swapStreamReference_; 
                }
                else
                {
                      return this.swapStreamReference_; 
                }
            }
            set
            {
                if (this.swapStreamReference_ != value)
                {
                    this.swapStreamReference_ = value;
                }
            }
        }
        #endregion
        
        public string swapStreamReferenceIDRef_ { get; set; }
        #region BusinessDayConvention_
        private BusinessDayConventionEnum businessDayConvention_;
        public BusinessDayConventionEnum BusinessDayConvention_
        {
            get
            {
                if (this.businessDayConvention_ != null)
                {
                    return this.businessDayConvention_; 
                }
                else if (this.businessDayConventionIDRef_ != null)
                {
                    businessDayConvention_ = IDManager.getID(businessDayConventionIDRef_) as BusinessDayConventionEnum;
                    return this.businessDayConvention_; 
                }
                else
                {
                      return this.businessDayConvention_; 
                }
            }
            set
            {
                if (this.businessDayConvention_ != value)
                {
                    this.businessDayConvention_ = value;
                }
            }
        }
        #endregion
        
        public string businessDayConventionIDRef_ { get; set; }
        
    
        
    
    }
    
}

